### DISCUSSION TITLE
Porting any of Fooocus-MRE features into original code base?

### DISCUSSION THREAD

**Discussion by MoonRide303** [lllyasviel/Fooocus#689] [Ideas]
I was maintaining [Fooocus-MRE](https://github.com/MoonRide303/Fooocus-MRE) for some time, and it got some fans (it was most popular fork until I archived the repo - over 500 stars), but having to re-write large parts of my code every time original Fooocus code base got bigger changes was too much for me - I was wasting a lot of my time on un-breaking things, instead of creating real value added. I guess that's the lesson about forking for me.

Most of the features I was missing in vanilla / original (like img2img, control over sampler / scheduler settings, or using recent ComfyUI backend) are already available in Fooocus v2.1 - and that's another reason I don't feel maintaining separate fork is that much needed.

I can still occasionally contribute to Fooocus, but I don't like wasting my time - so I'd rather contribute to original code base here, to avoid having to re-write same features multiple times.

So my question is: which - if any - of the [Fooocus-MRE features](https://github.com/MoonRide303/Fooocus-MRE#moonride-edition-features) could be a good fit for Fooocus? Some of those are already available in v2.1, but some are still available only in MRE (like [CLIP Skip control](https://github.com/MoonRide303/Fooocus-MRE/wiki/CLIP-Skip), [FreeU](https://github.com/MoonRide303/Fooocus-MRE/wiki/FreeU), and also small quality of life improvements - like support for JPEG, hotkey for starting generation, etc.). @lllyasviel please take a look on that list, and let me know if you'd like to see any of those features to be added into original code base - I can prepare pull requests adjusted to current Fooocus code base, if so.

**Comment by xhoxye** [comment]
1、Preset switching, anime with real and much more，On the UI interface
2、Starting generation via hotkey
3、Compact resolution and style selection，Search
4、Displaying full metadata for generated images in the UI.
5、Support for JPEG format.
6、Ability to save full metadata for generated images (as JSON or embedded in image, disabled by default).
7、Ability to load prompt information from JSON and image files (if saved with metadata).
8、History contains all and can be linked separately by date
9、custom styles and Preset 
Thanks.

**Reply by mirek190** [reply]
@MoonRide303  YES right now we have under fooous separate executables for anime and realistic presets ... that should be under fooocus itself option where we could choose preset next to "genetare" button (some list after click on it ? - but must be made with sense  ) plus custom presets that we could add.

**Comment by lllyasviel** [comment]
Clip skip will be somewhat complicated to implement because we use same vector for base and refiner ( even when refiner is 1.5, and this is one reason why we use 1.5 dreamsharper rather than 1.5 counterfeit as refiner for Fooocus anime, cus dreamsharper works better for final hidden layer ( and actually dreamsharper 8 is still marginally better than counterfeit for refining images even with clip skip -2 in my tests. Another method is loading refiner's clip to GPU but this is not acceptable in my opinion for speed and smoothness.

A more meaningful thing that should be considered before clip skip is changing vae if refiner is 1.5 (but most current 1.5 models has vae baked so .. ( but the baked vae in 1.5 counterfeit is pretty bad and this could be one reason to still have it 

Also keep in mind that the goal of Fooocus is to [explore new mediums of thought and expand the imaginative powers of the human species](https://www.midjourney.com/home/). This is the key that will determine whether the project Fooocus will be more and more popular, or just die like many other previous projects. Reconsider features if some features will limit people’s willingness to consider new ideas and preventing people from trying novel prompting and typing new descriptions. 

Most users of Fooocus are normal people who do not write reviews about software, just like that I never write a review after I buy a bicycle or a laptop. Some story-telling or documentary film makers (or some social media video makers), magazine editors, or say some blog writers are using Fooocus because they need images and they find this software, and they do not have Github account and their voices will hardly reach to you. The most difficult part is to know their needs, based on “ambient” feedbacks.

**Reply by MoonRide303** [reply]
Yeah, I understand the goal and philosophy behind Fooocus, and that any contributions should stay aligned to that. I had a bit different target user group in case of MRE, but I tried to stick to the same ground rules.

I also understand most of the users might be silent (no GitHub account, etc.), but listening to some of them might be still better option than listening to none - so I used feedback I had available, and filtered it through my own common sense. Not perfect (as I am pretty biased myself, far from being typical human species representative), but allows to keep feedback in the development loop in relatively easy way. Better solution would be probably working with experienced UX team, and letting them do the proper UX research - but that might be overkill for relatively small project like Fooocus :).

As for adding / porting specific features, we can talk about them in rounds - and nice to see FreeU already added, as well as support for embeddings :). As for other features:
1. CLIP Skip - changing it only for the base sounds good (impact for refiner is barely noticable).
2. Starting generation via CTRL+ENTER (ported from A1111) - doesn't take UI space, and it's pretty convenient. I never used it before, but after multiple users asked for it, and I've ported it and started using myself - it's now one of the first things I notice missing when I use vanilla :).
3. Compact Resolution list and support for custom resolutions - typing "1280x720" is way quicker than digging out through tabs and checkboxes to find separate "Forced Overwrite of Generating Width" and "Forced Overwrite of Generating Height".
4. Control-LoRA: Depth (rank 128 is good enough, and uses less VRAM) - it works pretty well without preprocessor and is able to produce nice results, IMO worth having (or any other depth CN, if you find it better).
5. Support for custom styles list (automatically imported from sdxl_styles/* on startup), possibly also including [sdxl_styles_mre.json](https://github.com/MoonRide303/Fooocus-MRE/blob/moonride-main/sdxl_styles/sdxl_styles_mre.json) (sample output [here](https://github.com/MoonRide303/Fooocus-MRE/wiki/Styles-(mre))).
6. Compact style selection - Dropdown also allows filtering, it's easier to find desired style(s) this way.

**Reply by lanyusan** [reply]
@MoonRide303 
Thanks a lot for your great effort in Fooocus-MRE. It was my main tool util now.

@lllyasviel 
On the note of making Fooocus a great tool for non-it guys, like most creative workers, I suggest to consider implementing this feature: pick an object from an image, and replace everything else. It is like reverse inpainting. Some products call it background replacement. But it should be more than simply background removal plus blending with new generated background. The object should be placed as a natural part of the new background. Simple cutout of object and placement in new background wouldn't make it.

It is very useful for product poster designers.

**Reply by killuams08** [reply]
I'm a 'silent' user, but I appreciate the effort very much.+1111111

**Reply by botzero-net** [reply]
It'd be great if we introduced standardized resolutions like 1920x1080.  I found another post here with a config change that allowed it and worked great but it gets overwritten with the original resolutions.

**Comment by TTTed** [comment]
My top 4 must-haves are: 
1. Control-LoRA
2. Full metadata log
3. Compact Resolution list
4. Custom style

**Reply by mirek190** [reply]
including that already realistic and anime ;)

**Comment by Pdonor** [comment]
Wildcards! I like being able to type __CustomPlaces__ and automatically grab from my list of places and so on. Thank you!

**Reply by MoonRide303** [reply]
@Pdonor Added in v2.1.687.

**Reply by MoonRide303** [reply]
@lllyasviel Could we get back to using seeded random for picking wildcards, instead of simple iterating? Not critical, but that change causes prompts using wildcards to yield different results in original, and MRE / Ruined.

**Reply by lllyasviel** [reply]
previous method sometimes give me same colors in a batch of 2. The change is 1/n where n is length of wild card.

**Reply by MoonRide303** [reply]
@lllyasviel That's how random functions should work: it allows duplicated choices in the series (like flipping coin twice with same result), and distribution will be going flat for larger number of samples.

As long as same seed yield same result it's okay, and your way will also cover all the choices - but for longer wildcards list it won't properly distribute choices over whole population. Imaging long list of alphabetically sorted artists, let's say 1000 of them. If you'd generate 10 images using `__artist__` wildcard your way might end up with generating only artist with names starting with letter A, instead of picking 10 random artist with different names from whole population.

**Reply by barepixels** [reply]
simple iterating is good for batch prompt list as some requested. maybe modify a copy of wildcards for "batch prompts"

**Comment by lllyasviel** [comment]
Hi everyone, we hear your voice and added notes to developer tools
![image](https://github.com/lllyasviel/Fooocus/assets/19834515/ab507d58-0db5-4615-b61d-3c8ce6558931)

**Reply by Aamir3d** [reply]
My suggestion here would be to implement the straightforward MRE interface flow - just typing in the resolution, or limiting the list of resolutions will make for a faster and more intuitive workflow.

**Reply by lllyasviel** [reply]
That makes Fooocus easy-to-fail rather than easy-to-use. Non-standard resolution leads to much more disordered arms and legs. See also SDXL tech report, page 17 at https://github.com/Stability-AI/generative-models/blob/57862fb4c76e72b0b867ebda3a435f0cefd71b4f/assets/sdxl_report.pdf

**Reply by Aamir3d** [reply]
I'm not disagreeing there @lllyasviel - the point is that the interface gets very cluttered from a UI perpsective.
The thing I found easier to use was this method implemented in MRE - 
1. Have dropdowns for all the regular resolutions that are SDXL approved
![image](https://github.com/lllyasviel/Fooocus/assets/122131887/bb9fbce2-d5f0-4193-a55b-08b4fb954682)
2. Type in resolution (divisible by 64) that you require - this is much faster and something I use constantly.
![image](https://github.com/lllyasviel/Fooocus/assets/122131887/d6597030-4e71-440e-a315-2443f2c8edc5)
3. Have a custom resolution list to reduce the clutter (make the interface more simple) 
Obviously this is for more advanced users, but it works well.

**Reply by MoonRide303** [reply]
@lllyasviel As long as the target resolution stays around 1 MP it works reasonably well:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/6efbe03c-0d1c-4e79-a441-be1ceed8bf50)

We could add a kind of safety net here, that is if user enters too low (or high) either resolution (let's say 0.5 MP to 2 MP) or aspect ratio (below 1:4 or above 4:1) then simply adjust it to nearest usable resolution.

**Reply by lllyasviel** [reply]
You should try some prompts with arms and legs.

"a woman sitting on grass", seed 0, standard 1152x896

all compostions are kind of acceptable in my opinion

![image](https://github.com/lllyasviel/Fooocus/assets/19834515/c592581d-77f6-4c39-b48a-1729cc76b7b4)

"a woman sitting on grass", seed 0, non-standard resolution 1280x720, note that SDXL is not trained on 1280x720

much more cropped contents, or back to viewer, or distortion

![image](https://github.com/lllyasviel/Fooocus/assets/19834515/52d43308-9241-46a0-9fa2-eb5cee124567)

**Reply by MoonRide303** [reply]
@lllyasviel Both sets stay within "acceptable" quality range for me, so I probably just allow it (like in MRE). I am also aware that using resolutions from the officially supported list will give best results, so pre-defined resolution list should stay as it is.

PS Another reason for allowing custom resolutions is support for SD base models, which give best results around 512x512. But support for SD and/or hybrid SD / SDXL workflows (SD base -> SDXL base + SDXL refiner, SD base -> SDXL refiner, etc.) is a bit broader topic, so we can leave if for another time (and using proper lower initial SD resolution could be done automatically in the background, if SDXL models would be used at later stages).

**Reply by lllyasviel** [reply]
do note that the compositions from 1280x720 is much more "cropped" like, feeling that they are images cropped from a bigger image. 

Also, if you allow custom resolutions, many non-geeks will really use more "common" numbers like 1000x1000 or 800x800, which makes results significantly worse in my tests.

**Reply by MoonRide303** [reply]
@lllyasviel Most casual users will probably just pick one of the predefined resolutions. And for the custom resolutions we could round it to nearest 16 / 32 / 64 px - it should solve 1000x1000 problem.

**Reply by lllyasviel** [reply]
Also I need to make sure that I am talking to a guy who knows how [positional encoding](https://www.google.com/search?q=positional+encoding&oq=positional+encoding&gs_lcrp=EgZjaHJvbWUyBggAEEUYOdIBBzM1OWowajGoAgCwAgA&sourceid=chrome&ie=UTF-8) works, otherwise I am wasting my time.

Basically, when a number, say “x”, is feed in to SDXL, the real vectors feed to model is not “x”. It is "f(x)" like

f(x)=torch.tensor([sin(x\*1), sin(x\*2), sin(x\*4), sin(x\*8), sin(x\*16), sin(x\*32), sin(x\*64), sin(x\*128), sin(x\*256), sin(x\*512)])

so the difference between 1024 and 1023 is not 1/1024. 

f(1024) and f(1023) are almostly 100% different.

**Comment by lllyasviel** [comment]
See also 
[Policy] About Forking and Promotion of Forks
 https://github.com/lllyasviel/Fooocus/discussions/699

**Comment by lllyasviel** [comment]
# About compact style selection

We are aware that some people like to have something like this:

![image](https://github.com/lllyasviel/Fooocus/assets/19834515/bccc69b0-472d-4338-9e9d-5b68a809cfa8)

This only needs to modify one line of code and we clearly have tried it.

The side effect is that new users are significantly less likely to try new styles. In an user study we did at the scale of 7 individuals before the Fooocus release, for current Fooocus UI (tested with 4 individuals, all first-time users), all 4 users try at least one styles and see what happens (1 user tried two styles and 1 user tried a lot). However, if the UI became like the above screenshot (tested with another 3 individuals, all first-time users), all 3 users ignore this selection (although 2 users still asked what is this, but 0 tried to change it). 

Although this design has some advantages, like easier removal of styles, the strong side effect prevents us from using it.

Also, you can search page with Ctrl+F.

Besides, loading style to prompt is not possible because Fooocus use neural network features to merge styles, rather than merging the texts.

**Reply by lllyasviel** [reply]
One possible explanation of this phenomenon (from my understanding) is that:

In the current UI, if a new user mistakenly turned a style off, one can click again at the exactly same position to turn that style on, this makes them “brave” enough to try edit it.

However, in the screenshot, if a style is removed, user has to remember the name of removed style to restore it (not sure if Ctrl+Z works but this does not seem reliable). So users are afraid of messing things and they are less likely to edit it.

**Reply by MoonRide303** [reply]
Compact style list with filtering allows you to quickly see all styles you might be interested in, like that:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/df032640-c7e1-4ab0-8f45-da640166cc65)

Which is (IMO) more convenient than searching for desired style using CTRL+F from browser. But that's just single and obviously biased user perspective ^^.

Merging the styles should work exactly the same way as before - it's only about how to present style selection in the UI.

As for the test on first-timers - question is what % of the first-time users should be assumed for that kind of tests. On one hand we want to have UI that's as much approachable as possible for new users, on the other - it should also be comfortable to use for people who already know the basics. Sadly those are sometimes contradictory goals, and it might be hard to balance it properly. It might be worth to perform tests on multiple groups with different levels of experience with computer graphics / generative art / Fooocus (first-timers, little experience, moderate experience, experts), and them aim for maximizing satisfaction of whole population (not just single group).

**Reply by xhoxye** [reply]
My suggestion is, why not show both? A search box plus the original.

**Comment by lllyasviel** [comment]
# About Custom Style

yes.

**Reply by MoonRide303** [reply]
Okay, I will merge it.

**Reply by lllyasviel** [reply]
if you merge, do test "run_anime.bat" and "run_realistic.bat" cus modifying styles have a chance to break them

**Reply by MoonRide303** [reply]
Tested, merged (https://github.com/lllyasviel/Fooocus/pull/704).

**Reply by mirek190** [reply]
how to load those custom styles now including anime and realistic?

**Reply by barepixels** [reply]
Tested, were able to to swap with my custom style list..  Thanks MoonRide303

mirek190, 
I noticed there are Anime and Realistic Presets, which you chose by launching bat file.  They use the config json files in the presets folder. 

 And there are Anime and Realistic Styles, which can be accessed in the checkboxes on the front end. They are in the `sdxl_styles_fooocus.json`  in the sdxl_styles folder.  These are accessible regardless which bat file you launch

Maybe in the future Fooocus can supports SD1.5 and 2.1 by having their own Presets/Bat files.  With their own user_path_config and resolution list. Call me crazy lolol

**Reply by mirek190** [reply]
I checked presets for anime and realistic from .bat files and from checkboxes .
Bat file works grat!
If I run normal  run.bat  and from checkbox choose Fooocus Anime or Fooocus Realistic then doesn't work like from the .bat files at all. 
For instance not changing model and lora  ... so are retarded ;)  actually... those presets not working properly at all ... not generating anime nor realistic pictures...   ( other presets works like SAI amine and others except  presets Fooocus-xxxx )

**Reply by MoonRide303** [reply]
@mirek190 Those presets are fine, but don't take them too literally. Those are just predefined Fooocus configurations allowing you to start Fooocus with different default settings - they can be only as "anime" or "realistic" as those settings allow it. You can check out files in `presets` folder to see the definitions.

**Reply by mirek190** [reply]
@MoonRide303 
So .... in my opinion they should not be there as not working as normal style modifier.... that is very confusing.

Maybe those buttons could be separated by a line or something and choosing one of them should change everything like .bat file are doing it?
Or even separate tab for them and extra custom configs.

**Reply by MoonRide303** [reply]
@mirek190 "Fooocus Realistic" and "Fooocus Anime" available in Style tab are just normal styles. They're being used in "anime" and "realistic" presets, but they're not otherwise connected. I understand it might be a bit confusing, but it's not really that much complicated :).

**Reply by mirek190** [reply]
@MoonRide303 From my test aren't normal as not changing style with standard sdxl model. 

Fooocus anime not making anime style also Fooocus realistic not making realistic pictures without those extra .bat files which making note changes.

**Reply by MoonRide303** [reply]
@mirek190 I've noticed @lllyasviel renamed "Fooocus Anime" to "Fooocus Masterpiece" - it should be less confusing, now.

@lllyasviel I have some doubts about default negative prompt in sdxl preset. It might be better to keep it empty (more generic) - reasons:
1. Previously prompt fields exposed to the end-users in the UI were more about high-level subject definition (what you want to see / what you don't want to see), and details related to quality and artistic style were handled at lower level (via styles and/or prompt expansion).
2. Quality-related negative prompts ("low quality", "ugly", "deformed", "bad anatomy", etc.) are usually included in the styles definition.
3. Phrases like "bad hands, bad eyes" etc. make sense when users ask for images representing (healthy) people and in styles related to that type of content (like human portraits), but might cause unwanted side-effects when trying to generate anything else (they can narrow models creativity).

**Reply by barepixels** [reply]
ARRRRRRRG  I can no longer have ONLY my custom style list in Style tab.  As of now every time I launch Fooocus, it auto download and auto load 5 default json and shove my custom to the very bottom.  I have to scroll and scroll past through styles that I do not want just to get to my favorites.  This is not an ELEGANCE design

Suggestions:

Auto download style.json to "sdxl_styles_library" folder and let users decide what to copy into "sdxl_styles" folder *****
and or
Refresh the loading style.json(s) similar to refresh models, post launch
and or
Load style json files by A-Z order.  sdxl_styles json(s) comes after favorite.json, photography.json, painterly.json etc..
and or
Different color blocks for each style.json, visual grouping
and or
Exclude list in setting from showing in Style Tab

**Reply by lllyasviel** [reply]
Now it is on top in latest version
![image](https://github.com/lllyasviel/Fooocus/assets/19834515/83b7be12-3509-47fe-ac00-fd6c100f6e7a)

**Reply by barepixels** [reply]
thank you

**Reply by MoonRide303** [reply]
@lllyasviel Another option - renaming defaut styles to something like `sdxl_styles_10_fooocus.json`, `sdxl_styles_20_sai.json` etc., and then treating all the files from `sdxl_styles` folder same way - users could then create whatever order they want (their styles as first, last, or in-between built-in default styles) just by using proper file names.

**Reply by barepixels** [reply]
as is now if you rename the default files, fooocus will download another sets of 5 default  json :(  correct?  I still think this is best:
Auto download style.json to "sdxl_styles_library" folder and let users decide what to copy/place into "sdxl_styles" folder
But I can live with current A-Z order.  Thanks again

**Reply by MoonRide303** [reply]
@barepixels Currently file names matter only for non built-in styles. If we dropped hard-coding file order in the code and used only file names for determining order of all the styles, then you could use something like `sdxl_styles_15_barepixels.json` to have your styles just after default ones, or `sdxl_styles_00_barepixels.json` to make them first on the list.

**Reply by barepixels** [reply]
i understand now.  thank you.  For now, I will just name all of my custom style with the beginning letter before "S"

**Reply by lllyasviel** [reply]
sdxl default negative is because I notice that some users like to disable all styles rightly after open Fooocus, so give a default negative is more flexible 

another reason is that I tested the `uc=clip_sdxl("")` and find that the vector `uc` is not close at all to the mean of possible conds, so the empty cond is more like a random cond with a random semantic that we cannot control. Having at least some texts in it so we al least know what is in the negative cond.

**Reply by lllyasviel** [reply]
@MoonRide303 For example, in SD1.5, the semantic of empty text "" is like this

(empty prompt, generated by A1111)
![image](https://github.com/lllyasviel/Fooocus/assets/19834515/aba0ad99-9371-4691-8369-af49a1a68679)

So in SD1.5, empty text "" is like "plain non-sense texture"

However, in XL, the semantic of empty text "" is like this

(disable all styles, prompt is "", no neg, seed is 0)
![image](https://github.com/lllyasviel/Fooocus/assets/19834515/1b670d02-0d23-46a5-a4e7-11849326b9e1)

so XL's "" does not mean nothing - it means generating a random object - and it is not bad enough as a negative

Below is the semantic of text "low quality, bad hands, bad eyes, cropped, missing fingers, extra digit"


(disable all styles, prompt is "low quality, bad hands, bad eyes, cropped, missing fingers, extra digit", seed is 0)
![image](https://github.com/lllyasviel/Fooocus/assets/19834515/71dd013f-69b3-4754-b686-87b2b7806075)

so it is as bad as 1.5's empty and more suitable for a negative

**Reply by lllyasviel** [reply]
If you want that level of customization you should directly use https://github.com/lllyasviel/Fooocus#customization and then probably copy/rename/write default styles in customized styles.
also sdxl default negative can be turned off in this config

**Reply by barepixels** [reply]
Ah thanks lllyasviel didn't notice this.  

```
    "default_negative_prompt": "low quality",
    "default_positive_prompt": "",
    "default_styles": [
        "Fooocus V2",
        "Default (Slightly Cinematic)",
        "SAI Photographic"
    ]
```
My bad.

**Reply by MoonRide303** [reply]
@lllyasviel Out of curiosity I also tested completely empty conditioning (zero-outed both positive and negative - Prompt Strength sliders in MRE), and... it seems SDXL model have seen a lot of cars - seed 0 to 7 all cars :).

As for the default negative prompt the way it works now just forces me to manually clear negative prompt each time I launch original, so I can start with clean environment - a bit annoying. People who would like to do raw prompting and starting from scratch will probably continue deselecting styles and prompt expansion as they did before, but now they also have to additionally clear negative prompt we're trying to enforce.

**Reply by lllyasviel** [reply]
i wll consider turn it off if in any example it leads to side effects.
but I tried many prompts and cannot find side effects

**Reply by lllyasviel** [reply]
also this is your prompt "perfect sportcar" seed 1

empty negative
![image](https://github.com/lllyasviel/Fooocus/assets/19834515/32aefab9-d42e-427e-aece-dfc79e4f850f)

current negative
![image](https://github.com/lllyasviel/Fooocus/assets/19834515/3e0f72f0-76c0-42ea-8d3c-6b0606c61175)

seems to help remove SDXL's blurry background problem

in most other cases effects are minor and relative random but it has a trend to help adress some Xl problems

**Reply by MoonRide303** [reply]
@lllyasviel My basic test case - "modern sports car", seed 0 (in v2.1.693).

With current defaults (negative prompt "low quality, bad hands, bad eyes, cropped, missing fingers, extra digit") I get something strange on the mask, road and the background are messed up:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/b8e851e8-5adf-4ccd-9c52-fcda20a49045)

vs clean negative prompt:
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/3e45ab99-2c3d-4634-bf97-1e9ff390d9ba)

**Reply by lllyasviel** [reply]
I do not think it really make sense to look too much at those seed-specific image details, but in most cases it can prevent background blur to some extent - that is already strong enough reason to have it

**Reply by lllyasviel** [reply]
also I forget to mention that in Fooocus, the negative prompt and all styles' negative texts are encoded by CLIP independently so that their texts do not join each other - whether to use each negative prompts' tokens depends on SDXL attention layer's softmax weight. 

so you do not need to worry about these texts are joining styles' negatives. If the model does not need it, it will just ignore it by using zeros in softmax.

**Comment by hzeasy** [comment]
Glad to express my opinion here, I think "Save Metadata in Image" and "load prompt" are great features,Hope to join fooocus。Thanks！
![4a5954318e3c87d3f56921ae0921caf](https://github.com/lllyasviel/Fooocus/assets/136664755/503c24e7-0839-41cb-b0d2-ef3ddc3a485e)
![7891e99ccad7317e2cc61fd22c8e7c9](https://github.com/lllyasviel/Fooocus/assets/136664755/90491ac6-7eb3-4c24-9276-47996f012504)

**Comment by AkoZ** [comment]
ouchhh ! i discover lately about even Fooocus (1 week now and really enjoying it) and the MRE one which seems a bit more dev adapted, i think about 2 kind generous persons able to get the best parts of their own engines: the comfy inside, which has some of the needed base modules here added, as metadata inside .png and the way easy to get a whole workflow in one .png deposit is so crazy cool !
* easy i just suppose this could be fast implemented in fooocus but (? well who knows enough) - as i said on https://github.com/lllyasviel/Fooocus/discussions/552 there are good parts in comfy not kept in fooocus ? 
* 
Also is the ruinedFooocus too ! each has their goodies, but well philosophies too..
But hey is the Python_embeded (about 5Go) a shared/the samle for the threes Fooocus ?
If so plzz make 
**a separate principal Fooocus child dir** to minimize the datas but also to have the fast option for the day.
*
as to have the most of the 2: **MRE inside** have those good things here :
1. Support for embeddings (use "embedding:embedding_name" syntax, ComfyUI style).
2. Displaying full metadata for generated images in the UI.
3. Support for JPEG format.
4. Ability to save full metadata for generated images (as JSON or embedded in image, disabled by default).
5. Ability to load prompt information from JSON and image files (if saved with metadata).
6. Ability to generate multiple images using same seed sure and DONE, then there is somewhere like in comfy a just "+1" seed, easy to get a suite, also a bit strange for the names of images in day_directory to have random numbers, but easier +1 too ;)
7. **DONE yeaahhhh** : sure, add **MRE custom styles** and 
8. **Have the styles in "ontology" classed** (as it begins to go too far down to get the whole thing), may be by some hidden/open dirs/tabs/accordion - yep 2nd would be image inside (hey take a look at [easy diffusion ](https://github.com/easydiffusion/easydiffusion/wiki/Image-Modifiers) )
9. As Easy diff with **Seamless Tiling on image** .. so good
10. Style Iterator (iterates over selected style(s) combined with remaining styles - S1, S1 + S2, S1 + S3, S1 + S4, and so on; for comparing styles pick no initial style, and use same seed for all images).

**Cheers the devs !** [MoonRide303](https://github.com/MoonRide303) and [lllyasviel ](https://github.com/lllyasviel) and [runew0lf](https://github.com/runew0lf/RuinedFooocus/commits?author=runew0lf) too...

**Reply by runew0lf** [reply]
<3

**Reply by AkoZ** [reply]
there is a fabulous thing with ComfyUI: it can replace the whole workflow just by putting an image on screen !
* Really (here the 5-), this would be so crazy easy to get the whole config from inside a .PNG directly...
* ok i've seen sthing about noway to let any "prompting" inside the image as lost/stolen data:
* BUT for sure, as ethical as it is .. there can be a user choice option: keep data in .png or not ?
* OR may be .. as directories are used (i rename fast it when i think i will work on an other thema image) .. and that's it, really enjoyable to use, and for sure my near last images get the right prompt found from step by step and essays: 
* 
* so **just KEEP ONE "Starter image"** inside this directory which contains the whole config (a .json ? also ok): 
* but finding back the whole conf, as for now .. hey i'm on sometimes screen capt the dev conf (even the FreeU)...

An other starred option is to **put the styles from hidden/open tabs**, as MRE integrated ! (yyeahahh) and "artstyles", those are really in categories (and then no way to put the MRE or artstyle on every button- this would shorten the long search for the lost styles
 _...what a long start to re-config back and re-finding the whole things ! really annoying..._
 * hey i don't cry ! .. you're so societal DEVs ! Precursors of the next Earthians Nation ;)

**Reply by AkoZ** [reply]
@runew0lf May be as i review your [Version as RuinedFooocus](https://github.com/runew0lf/RuinedFooocus/tree/main), you have integrated many of those asks ? - _sshooww , an other foocus to install, even easy... but once more to install._

1.  is there a way to (yeah so magically) pass from "standard" FoOocus to Ruined ? As to replace some files but keep the whole
2.  OR is it fundamentally different directories organisation and even the python_embeded dir is also crashed to an other ?

**Reply by runew0lf** [reply]
Its a lot different nowadays, each of us had our own vision of where it should be going!
